File: typescript/scheme/scheme_expressive_test.ts
/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'jasmine';

import {MaterialDynamicColors} from '../dynamiccolor/material_dynamic_colors.js';
import {Hct} from '../hct/hct.js';
import {customMatchers} from '../utils/test_utils.js';

import {SchemeExpressive} from './scheme_expressive.js';

beforeEach(() => {
  jasmine.addMatchers(customMatchers);
});

describe('scheme expressive test', () => {
  it('keyColors', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 0.0);

    expect(MaterialDynamicColors.primaryPaletteKeyColor.getArgb(scheme))
        .matchesColor(0xff35855F);
    expect(MaterialDynamicColors.secondaryPaletteKeyColor.getArgb(scheme))
        .matchesColor(0xff8C6D8C);
    expect(MaterialDynamicColors.tertiaryPaletteKeyColor.getArgb(scheme))
        .matchesColor(0xff806EA1);
    expect(MaterialDynamicColors.neutralPaletteKeyColor.getArgb(scheme))
        .matchesColor(0xff79757F);
    expect(MaterialDynamicColors.neutralVariantPaletteKeyColor.getArgb(scheme))
        .matchesColor(0xff7A7585);
  });

  it('lightTheme_minContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xff32835D);
  });

  it('lightTheme_standardContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xff146C48);
  });

  it('lightTheme_maxContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xff002818);
  });

  it('lightTheme_minContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xffA2F4C6);
  });

  it('lightTheme_standardContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xffA2F4C6);
  });

  it('lightTheme_maxContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xff004D31);
  });

  it('lightTheme_minContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xff1e724e);
  });

  it('lightTheme_standardContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xff002112);
  });

  it('lightTheme_maxContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xff9aebbe);
  });

  it('lightTheme_minContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, -1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xfffdf7ff);
  });

  it('lightTheme_standardContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 0.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xfffdf7ff);
  });

  it('lightTheme_maxContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), false, 1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xfffdf7ff);
  });

  it('darkTheme_minContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xff32835d);
  });

  it('darkTheme_standardContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xff87d7ab);
  });

  it('darkTheme_maxContrast_primary', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.primary.getArgb(scheme))
        .matchesColor(0xffd5ffe4);
  });

  it('darkTheme_minContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xff005234);
  });

  it('darkTheme_standardContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xff005234);
  });

  it('darkTheme_maxContrast_primaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.primaryContainer.getArgb(scheme))
        .matchesColor(0xff8bdbaf);
  });

  it('darkTheme_minContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xff76c59b);
  });

  it('darkTheme_standardContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xffa2f4c6);
  });

  it('darkTheme_maxContrast_onPrimaryContainer', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.onPrimaryContainer.getArgb(scheme))
        .matchesColor(0xff004229);
  });

  it('darkTheme_minContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, -1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xff14121a);
  });

  it('darkTheme_standardContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 0.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xff14121a);
  });

  it('darkTheme_maxContrast_surface', () => {
    const scheme = new SchemeExpressive(Hct.fromInt(0xff0000ff), true, 1.0);
    expect(MaterialDynamicColors.surface.getArgb(scheme))
        .matchesColor(0xff14121a);
  });
});

File: cpp/scheme/scheme_expressive_test.cc
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme_expressive.h"

#include "testing/base/public/gunit.h"
#include "cpp/cam/hct.h"
#include "cpp/dynamiccolor/material_dynamic_colors.h"

namespace material_color_utilities {

namespace {
TEST(SchemeExpressiveTest, KeyColors) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryPaletteKeyColor().GetArgb(scheme),
            0xff35855f);
  EXPECT_EQ(MaterialDynamicColors::SecondaryPaletteKeyColor().GetArgb(scheme),
            0xff8c6d8c);
  EXPECT_EQ(MaterialDynamicColors::TertiaryPaletteKeyColor().GetArgb(scheme),
            0xff806ea1);
  EXPECT_EQ(MaterialDynamicColors::NeutralPaletteKeyColor().GetArgb(scheme),
            0xff79757f);
  EXPECT_EQ(
      MaterialDynamicColors::NeutralVariantPaletteKeyColor().GetArgb(scheme),
      0xff7a7585);
}

TEST(SchemeExpressiveTest, lightTheme_minContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff32835d);
}

TEST(SchemeExpressiveTest, lightTheme_standardContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff146c48);
}

TEST(SchemeExpressiveTest, lightTheme_maxContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff002818);
}

TEST(SchemeExpressiveTest, lightTheme_minContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff99eabd);
}

TEST(SchemeExpressiveTest, lightTheme_standardContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xffa2f4c6);
}

TEST(SchemeExpressiveTest, lightTheme_maxContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff004d31);
}

TEST(SchemeExpressiveTest, lightTheme_minContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff146c48);
}

TEST(SchemeExpressiveTest, lightTheme_standardContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff002112);
}

TEST(SchemeExpressiveTest, lightTheme_maxContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xffffffff);
}

TEST(SchemeExpressiveTest, lightTheme_minContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffdf7ff);
}

TEST(SchemeExpressiveTest, lightTheme_standardContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffdf7ff);
}

TEST(SchemeExpressiveTest, lightTheme_maxContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), false, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xfffdf7ff);
}

TEST(SchemeExpressiveTest, darkTheme_minContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff51a078);
}

TEST(SchemeExpressiveTest, darkTheme_standardContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xff87d7ab);
}

TEST(SchemeExpressiveTest, darkTheme_maxContrast_primary) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Primary().GetArgb(scheme), 0xffeefff2);
}

TEST(SchemeExpressiveTest, darkTheme_minContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff00432a);
}

TEST(SchemeExpressiveTest, darkTheme_standardContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff005234);
}

TEST(SchemeExpressiveTest, darkTheme_maxContrast_primaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::PrimaryContainer().GetArgb(scheme),
            0xff8bdbaf);
}

TEST(SchemeExpressiveTest, darkTheme_minContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff64b48a);
}

TEST(SchemeExpressiveTest, darkTheme_standardContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xffa2f4c6);
}

TEST(SchemeExpressiveTest, darkTheme_maxContrast_onPrimaryContainer) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::OnPrimaryContainer().GetArgb(scheme),
            0xff000000);
}

TEST(SchemeExpressiveTest, darkTheme_minContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, -1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff14121a);
}

TEST(SchemeExpressiveTest, darkTheme_standardContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 0.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff14121a);
}

TEST(SchemeExpressiveTest, darkTheme_maxContrast_surface) {
  SchemeExpressive scheme = SchemeExpressive(Hct(0xff0000ff), true, 1.0);
  EXPECT_EQ(MaterialDynamicColors::Surface().GetArgb(scheme), 0xff14121a);
}
}  // namespace
}  // namespace material_color_utilities
